src/pybind/mgr/volumes/fs/operations/versions/metadata_manager.py

```diff
@@ -42,7 +46,14 @@ def refresh(self):
         conf_data = StringIO()
         try:
             log.debug("opening config {0}".format(self.config_path))
-            fd = self.fs.open(self.config_path, os.O_RDONLY)
+            target_path = self.config_path
```

Comment:
I am leaning towards not maintaining a per incarnation .meta.<UUID> file. The information can be retained in the .meta file as such, and snapshots would retain the .meta state as when it was taken, to reference the incarnation ID.

Addition of the symlink only causes some twisted code to maintain the same.

Reply:
> Couple of points (and questions):
> 
>     * How would (and where) clone state (clone record in parent subvolume) be recorded if per-incarnation metadata is not maintained?

(also answering/addressing/pondering the bullets further below in the comment)

It is maintained in the .meta file, under the section "clone snaps" as before. The data in this section already encodes the (snapname, track-id) tuple, which gives the required path to access the .meta file within the snapshot to determine the incarnation the snapshot belongs to.

**NOTE:** The one place where we need to construct the snapshot incarnation directory to clone from, is handled in the function `subvolume_snapshot_path` which uses the .meta file within the snapshot to determine its subvolume incarnation path.

On a `rm` with `--retain-snaps` the .meta file global section is updated, as needed, to reflect that there is no current incarnation (I am being purposefully vague here about how the global section is updated to reflect the same, as we could use state/path/incarnation to do this, as is being discussed in other comments). All other sections in the .meta file are retained, including "protected snaps" and "clone snaps" sections.

When the last snapshot of a subvolume with no current incarnation is deleted, the subvolume is also deleted. (this is not coded yet)

My question/confusion is why do we need the "clone snaps" section recorded per-incarnation? I fail to understand that.

> I am not _really_ against keeping a single metadata instance as long as:
> 
>     * subvolume incarnations can be isolated to the versions they were created
> 
>     * can handle a mix of subvolume incarnations with and without `--retain-snapshots`
> 
>     * clone records can persist across subvolume incarnations
> 
> 
> Having symbolic links is a side-effect of using filesystem scratch space for metadata and to speed-up lookup operations for current incarnation.

Reply:
> My question/confusion is why do we need the "clone snaps" section recorded per-incarnation? I fail to understand that.
> 
> >

Technically, right now, there is no hard requirement for 'clone snaps' section to be recorded in per-incarnation metadata -- or any other metadata for that matter -- once a subvolume is deleted, metadata that needs to be recorded for that incarnation can be stashed in one single .meta (shared section in metadata file). Will it work -- sure it will. Is it a clean design -- I don't think so -- because the (deleted) subvolume has certain parts of its metadata (read-only) in the snapshot and other metadata (read+write) outside the snapshot. Also, what if the (writable) metadata is subvolume version dependent? One can figure out the version from the metadata in the snapshot. However, when we move towards storing subvolume metadata in a key/value store, there won't be any metadata in the snapshot to figure out which version a (deleted) subvolume got created.

Reply:
> (follow up question)
> 
> How does an incarnated subvolume relate its metadata when writable metadata (post subvolume deletion) is stashed in a place that's shared with other incarnations (including current, if any)?

Incarnations are properties of the same subvoume, the subvolume is the primary construct. Hence subvolume metadata having information regarding different incarnations is expected and not a violation.

> Right now, there is a 1:1 map between a subvolume version and its metadata, i.e., a given version is fully capable of understanding its metadata format. Sharing metadata pieces between incarnations seems to introduce unnecessary complications. E.g., say, a snapshot exist for subvolume incarnation 1 with version 2 and the current incarnation is 10 with version 5 -- stashing version 2 metadata in version 5 format does not look clean. Moreover, how does all of this tie-up with implementation? Should a version 2 subvolume _know_ how to persist metadata in (possibly _all_) other versions, or, should a version 5 subvolume _know_ how to deal with version 2 metadata? Right now, the metadata we are discussing is generic across versions. However, this _might_ not be the case going forward. In short, the question that we need to ask ourselves is: do we see a need to have metadata that is version dependent going forward?

Incarnations are not different versions. The version is tied to the subvolume, hence I do not see the above problem occurring.

If a version-n subvolume is retained, then its metadata and other on disk representations need to be upgraded to version-n++. Till such a time it remains as a version-n subvolume, and when upgraded it will not have any version-n subparts in its metadata.

Under these conditions, it is not feasible for a version-5 subvolume to have version-2 subparts in it, or a version-2 subvolume to have a version-5 incarnation created within it.

> 
> Also, how does all of this work with snapshot restore operation that will come up sooner or later? Does persisting both "clone source" and "clone record" in the same metadata with possibly different versions sound comprehensible? And what if the restore operation fails? Cleaning that up (removing half-baked subvolume) will not be a plain asynchronous purge.

Incarnations are not different versions, as laid out above.

> 
> IMO, this is leading to a considerable amount of complexity for a small win in dealing with a single instance (and possibly bloated) metadata chunk.

I think there is a gap here regarding how we perceive retained subvolumes. IMO, A retained subvolume that is recreated in the future is the same subvolume, hence its version/metadata are properties of the subvolume and not of its incarnations. Incarnations are way to separate retained-recreated chain of events on such a subvolme.

Reply:
> Incarnations are not different versions. The version is tied to the subvolume, hence I do not see the above problem occurring.
> 
> If a version-n subvolume is retained, then its metadata and other on disk representations need to be upgraded to version-n++. Till such a time it remains as a version-n subvolume, and when upgraded it will not have any version-n subparts in its metadata.

mgr/volumes provisions subvolumes with the highest subvolume version it understands. A subvolume being retained can be version X and then recreated with version Y.

Reply:
.. and BTW, who and when does the version "upgrade"?

Reply:
> > Incarnations are not different versions. The version is tied to the subvolume, hence I do not see the above problem occurring.
> > If a version-n subvolume is retained, then its metadata and other on disk representations need to be upgraded to version-n++. Till such a time it remains as a version-n subvolume, and when upgraded it will not have any version-n subparts in its metadata.
> 
> mgr/volumes provisions subvolumes with the highest subvolume version it understands. A subvolume being retained can be version X and then recreated with version Y.

Good catch, this has to be fixed (in this PR) to use the current version subvolumes create/create_clone method (if a current version for the subvolume exists).

Reply:
> .. and BTW, who and when does the version "upgrade"?

I would expect an auto-upgrade when possible (like legacy to v1, or v1 without snapshots to v2), as the first method.

Where auto upgrade to a particular future version is not possible, the user would need to upgrade the same (using any provided steps/method for the upgrade).

Reply:
Why not allow (re)creating subvolumes with the current version? Is that something specific to CSI?

I think allowing subvolumes to be recreated with current (available) version allows for an easier "upgrade" path -- take a snapshot, restore the subvolume from the snapshot (which is a rm+clone)

Reply:
> I would expect an auto-upgrade when possible (like legacy to v1, or v1 without snapshots to v2), as the first method.

The auto-upgrade path in mgr/volumes is for existing subvolumes to use features such as clone. We should generally avoid auto-uprgade on access kind of things (N-1 -> N or worse N-2 -> N). Moreover, IMO the recommended way to upgrade would be snap+restore. That requires maintaining multiple versions of a subvolume during upgrade, right?